File: hs/Day01.hs
module Day01 (day01, part1, part2) where

import Text.Printf

day01 :: IO ()
day01 = do
  input <- readFile "input/input01.txt"
  let dirs = map parse $ lines input
  printf "part 1 = %d\n" $ part1 dirs
  printf "part 2 = %d\n" $ part2 dirs

parse :: String -> Int
parse ('L' : num) = -read num
parse ('R' : num) = read num
parse _ = error "parse error"

part1 :: [Int] -> Int
part1 input =
  part1' input 50
  where
    part1' (d : ds) curr
      | curr' == 0 = 1 + part1' ds curr'
      | otherwise = part1' ds curr'
      where
        curr' = (curr + d) `mod` 100
    part1' [] _ = 0

part2 :: [Int] -> Int
part2 input =
  part2' input 50
  where
    step dist pos
      | dist == 0 = 0
      | otherwise = c + step (dist - sign) (pos + sign)
      where
        c = if pos `mod` 100 == 0 then 1 else 0
        sign = signum dist
    part2' (d : ds) curr = step d curr + part2' ds ((curr + d) `mod` 100)
    part2' [] _ = 0

File: src/bin/day01.rs
fn parse(input: &str) -> Vec<i64> {
    input
        .lines()
        .map(|s| {
            if let Some(num) = s.strip_prefix("L") {
                -num.parse::<i64>().unwrap()
            } else if let Some(num) = s.strip_prefix("R") {
                num.parse().unwrap()
            } else {
                0
            }
        })
        .collect()
}

fn part1(puzzle: &[i64]) -> i64 {
    let mut acc = 50;
    let mut count = 0;

    for dir in puzzle {
        acc += dir;
        acc %= 100;

        if acc == 0 {
            count += 1;
        }
    }

    count
}

fn part2(puzzle: &[i64]) -> i64 {
    let mut acc = 50;
    let mut count = 0;

    for dir in puzzle {
        let steps = dir.abs();
        let sign = if *dir > 0 { 1 } else { -1 };

        for _ in 0..steps {
            acc += sign;
            acc %= 100;

            if acc == 0 {
                count += 1;
            }
        }
    }

    count
}

fn main() {
    let input = include_str!("../../input/input01.txt");
    let input = parse(input);
    println!("part1 = {}", part1(&input));
    println!("part1 = {}", part2(&input));
}

#[test]
fn test_day01() {
    let input = "\
L68
L30
R48
L5
R60
L55
L1
L99
R14
L82";
    let input = parse(input);

    assert_eq!(part1(&input), 3);
    assert_eq!(part2(&input), 6);
}
